File: deed-vectorz/src/deed/vectorz.clj
(ns deed.vectorz
  (:import
   (deed Encoder
         Decoder
         OID)
   (mikera.vectorz IVector
                   Vectorz
                   AVector))
  (:require
   [deed.core :as d]))

(set! *warn-on-reflection* true)


(d/expand-encode [OID/MIKERA_IVECTOR ;; TODO: AVECTOR
                  AVector
                  iv
                  ^Encoder encoder]
  (let [len (.length iv)]
    (.writeInt encoder len)
    (loop [i 0]
      (if (= i len)
        nil
        (let [x (.get iv i)]
          (.writeDouble encoder x)
          (recur (inc i)))))))


(d/expand-decode [OID/MIKERA_IVECTOR
                  ^Decoder decoder]
  (let [len (.readInteger decoder)
        res (Vectorz/newVector len)]
    (loop [i 0]
      (if (= i len)
        res
        (let [x (.readDouble decoder)]
          (.set res i x)
          (recur (inc i)))))))

File: deed-vectorz/src/java/deed/VectorZ.java
package deed;

import mikera.vectorz.*;

public class VectorZ {

    public void encodeIVector(final Encoder encoder, final short oid, final IVector iv) {
        encoder.writeOID(oid);
        final int len = iv.length();
        encoder.writeInt(len);
        for (int i = 0; i < len; i++) {
            encoder.writeDouble(iv.get(i));
        }
    }

    public void encodeVector(final Encoder encoder, final mikera.vectorz.Vector v) {
        encodeIVector(encoder, OID.MIKERA_VECTOR, v);
    }

    public void encodeVector1(final Encoder encoder, final mikera.vectorz.Vector1 v1) {
        encodeIVector(encoder, OID.MIKERA_VECTOR1, v1);
    }

    public void encodeVector2(final Encoder encoder, final mikera.vectorz.Vector2 v2) {
        encodeIVector(encoder, OID.MIKERA_VECTOR2, v2);
    }

    public void encodeVector3(final Encoder encoder, final mikera.vectorz.Vector3 v3) {
        encodeIVector(encoder, OID.MIKERA_VECTOR3, v3);
    }

    public void encodeVector4(final Encoder encoder, final mikera.vectorz.Vector4 v4) {
        encodeIVector(encoder, OID.MIKERA_VECTOR4, v4);
    }

    public IVector decodeVector(final Decoder decoder) {
        final int len = decoder.readInteger();
        final IVector vector = Vectorz.newVector(len);
        double x;
        for (int i = 0; i < len; i++) {
            x = decoder.readDouble();
            vector.set(i, x);
        }
        return vector;
    }

}
